components/butterfly/src/member.rs

```diff
@@ -495,130 +489,58 @@ impl MemberList {
     ///
     // TODO (CM): why don't we just insert a membership record here?
     pub fn insert(&self, incoming_member: Member, incoming_health: Health) -> bool {
-        let accept_and_propagate_rumor = match self.members_read().get(&incoming_member.id) {
-            None => true,
-            Some(existing_member) => {
-                match existing_member
-                    .incarnation
-                    .cmp(&incoming_member.incarnation)
+        // Is this clone necessary, or can a key be a reference to a field contained in the value?
+        // Maybe the members we store should not contain the ID to reduce the duplication?
+        let modified = match self.members_write().entry(incoming_member.id.clone()) {
+            hash_map::Entry::Occupied(mut entry) => {
+                let mut val = entry.get_mut();
+                // maybe break this condition out?
+                if incoming_member.incarnation > val.0.incarnation
+                    || (incoming_member.incarnation == val.0.incarnation && incoming_health > val.1)
                 {
-                    cmp::Ordering::Greater => false,
-                    cmp::Ordering::Less => true,
-                    cmp::Ordering::Equal => {
-                        // We know we have a health if we have a
-                        // record.
-                        //
-                        // (Doing this in two operations is necessary
-                        // due to the lock and lifetimes.)
-                        let health_lock = self.health.read().expect("Health lock is poisoned");
-                        let existing_health = health_lock.get(&existing_member.id).expect(
-                            "No health for a membership record should be impossible; did you use insert?",
-                        );
-
-                        incoming_health > *existing_health
-                    }
+                    *val = (incoming_member, incoming_health, SteadyTime::now());
+                    true
+                } else {
+                    false
                 }
             }
+            hash_map::Entry::Vacant(entry) => {
+                entry.insert((incoming_member, incoming_health, SteadyTime::now()));
+                true
+            }
```

Comment:
It *almost* works. Here's what I tried:
```rust
        let mut modified = true;
        self.write_entries()
            .entry(incoming.member.id.clone())
            .and_modify(|val| {
                if incoming.newer_or_less_healthy_than(val.member.incarnation, val.health) {
                    *val = member_list::Entry {
                        member: incoming.member,
                        health: incoming.health,
                        health_updated_at: SteadyTime::now(),
                    };
                } else {
                    modified = false
                }
            })
            .or_insert(member_list::Entry {
                member: incoming.member,
                health: incoming.health,
                health_updated_at: SteadyTime::now(),
            });
```

The problem is that because `and_modify` takes a closure, the `incoming` values are moved into it and then can't be used by the `or_insert` case.

Additionally, the logic for setting `modified` ends up using a mutable variable which I'm less a fan of. To me, it's too easy to change one of the cases and forget to make the appropriate update to `modified`. As opposed to the existing implementation where all code paths return a `bool`, making it harder to forget.